### PR TITLE
Recsys losses

### PR DESCRIPTION
**Pull request by sergunya17**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [x] Did you run [colab minimal CI/CD](https://colab.research.google.com/github/catalyst-team/catalyst/blob/master/examples/notebooks/colab_ci_cd.ipynb) with `latest` and `minimal` requirements?

## Description

Added losses:
- AdaptiveHingeLoss
- BPRLoss
- HingeLoss
- LogisticLoss
- WARPLoss

Based on #1253, #1255

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: catalyst/contrib/nn/criterion/recsys.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            negative_scores: torch.Tensor
+                Iterable of tensors containing predictions for sampled negative items.
+                More tensors increase the likelihood of finding ranking-violating
+                pairs, but risk overfitting.
+        """
+        self._assert_equal_size(positive_score, negative_scores[0])
+
+        highest_negative_score, _ = torch.max(negative_scores, 0)
+
+        return HingeLoss().forward(positive_score, highest_negative_score.squeeze())
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/recsys.py]: could you create `HingeLoss()` once during `__init__`?

File: catalyst/contrib/nn/criterion/recsys.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    Pointwise approaches look at a single document at a time in the loss function.
+    For a single documents predict it relevance to the query in time.
+    The score is independent for the order of the documents that are in the query's results.
+
+    Input space: single document d1
+    Output space: scores or relevant classes
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/recsys.py]: do we need it? 👀

File: catalyst/contrib/nn/criterion/recsys.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    .. _WSABIE: Scaling Up To Large Vocabulary Image Annotation:
+        https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/37180.pdf
+
+    WARP loss randomly sample output labels of a model, until it finds a pair
+    which it knows are wrongly labelled and will then only apply an update to
+    these two incorrectly labelled examples.
+    """
+
+    def __init__(self, max_num_trials: Optional[int] = None):
+        super(WARPLoss, self).__init__()
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/recsys.py]:
```suggestion
        super().__init__()
```

**Review by Scitator — COMMENTED** [review]
a few questions here:
1. do we really need `Pointwise`, `PairwiseLoss` and `ListWiseLoss` abstractions?
2. could you please add usage examples for each implemented criterion? I hope, it would be much easier to understand them in this way :)

File: catalyst/contrib/nn/criterion/recsys.py (hunk 2 above)
Comment (sergunya17) [reply catalyst/contrib/nn/criterion/recsys.py]: 👀

**Pull request merged**